Bump version to 6.0.1 on the 6.x branch

DaveCTurner opened the pull request:
Mainly opening this PR to get a full CI run before merging, as it's mostly a port of #27386 to the 6.x branch. The changes to `VersionUtils#resolveReleasedVersions()` and associated tests deserve attention, and ultimately backporting to 6.0 too.

I'd wait for CI to pass before looking at this too hard.

DaveCTurner committed:
Bump version to 6.0.1

DaveCTurner committed:
Prepare for bump to 6.0.1 on the 6.0 branch (#27386)

An attempt to bump to 6.0.1 on the 6.0 branch exposed up a handful of issues that this commit fixes. One of those fixes is a terrible hack that will be fixed more thoroughly in #27397, and another is a back port of d5e56c55553291682932d7e9e8dc7068f59e618b which is related to #27251.

DaveCTurner committed:
Remove big horrible hack

Version.CURRENT.minimumCompatibilityVersion() is now Version.V_5_6_0 anyway.

DaveCTurner committed:
Fix assertion message

The preceding line modifies `versions` so `versions.get(versions.size() - 1)`
no longer contains the highest version.

DaveCTurner committed:
Rewrite the guts of VersionUtils.resolveReleasedVersions()

The policy for which version constants are unreleased changed at 5.6 - we now
plan to keep an unreleased 'marker' version at the end of each minor branch,
whereas prior to 5.6 the unreleased marker version was removed during the
version bumping.

This commit simplifies the logic a bit and also makes it clear which bits can
be removed once 5.x versions are no longer relevant.

DaveCTurner committed:
orly

DaveCTurner committed:
Improve comments

DaveCTurner committed:
Even better comments

DaveCTurner commented:
CI passed, so a review of my changes to `VersionUtils` would be appreciated.

s1monw commented on core/src/test/java/org/elasticsearch/VersionTests.java:
```diff
@@ -343,8 +343,6 @@ public void testIsCompatible() {
         } else {
             assertFalse(isCompatible(Version.CURRENT, Version.fromString("7.0.0")));
         }
-        assertFalse("only compatible with the latest minor",
```
why do we remove this assertion, I think it's valid?

DaveCTurner replied:
Sorry, yes, you're right it's valid. There's a bug somewhere. Hunting it now...

s1monw reviewed (changes requested):
left one question LGTM otherwise

DaveCTurner committed:
rly!

Actually this shouldn't fail on the 6.x branch - it's asserting that 7.0.0 is
compatible with 6.1.0 but not with anything on the 6.0.x series.

DaveCTurner committed:
Extract method

DaveCTurner committed:
Move reflection method to Version class

DaveCTurner committed:
Version#minimumCompatibilityVersion() now searches for the last minor of the previous major

DaveCTurner committed:
This assertion is the wrong one

DaveCTurner committed:
No need for special case for 5.6<->6.x any more

DaveCTurner committed:
Add javadoc

DaveCTurner commented:
@simonw ok, assertion reinstated and fixed with a change to `Version.minimumCompatibilityVersion`, and the conflicting `assertTrue(isCompatible(Version.fromString("6.0.0"), Version.fromString("7.0.0")));` removed instead. Could you take another look?

DaveCTurner committed:
Fix comment & extra whitespace

s1monw reviewed (approved):
LGTM

DaveCTurner committed:
Make things final

DaveCTurner committed:
Test classes need to be public for reflection to work

DaveCTurner committed:
Revert "Bump version to 6.0.1" - will be committed separately

This reverts commit a46d44ec56a23dfad16a221ce6b945134ba27bac.

jasontedor reviewed (commented):
LGTM.
